File: sources/scalac/util/FreshNameCreator.java
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
\*                                                                      */

// $Id$

package scalac.util;

import java.util.HashMap;

public class FreshNameCreator {

    protected int counter = 0;
    protected HashMap counters = new HashMap();

    /**
     * Create a fresh name with the given prefix. It is guaranteed
     * that the returned name has never been returned by a previous
     * call to this function with the same separator character (which
     * has to be a non-digit).
     */
    public Name newName(String prefix, char separator) {
        prefix += separator;
        Integer ival = (Integer)counters.get(prefix);
        if (ival == null)
            counters.put(prefix, ival = new Integer(0));
        else
            counters.put(prefix, ival = new Integer(ival.intValue() + 1));
        return Name.fromString(prefix + ival);
    }

    /** Same, with `$' as the separator character
     */
    public Name newName(String prefix) {
        return newName(prefix, '$');
    }

    /** Same, but with a name as prefix. The new name is a type
     *  (respectively, constructor) name if the prefix is one.
     */
    public Name newName(Name prefixName, char separator) {
	Name name = newName(prefixName.toString(), separator);
	if (prefixName.isTypeName()) return name.toTypeName();
	else if (prefixName.isConstrName()) return name.toConstrName();
	else return name;
    }

    /** Same, with `$' as the separator character
     */
    public Name newName(Name prefix) {
        return newName(prefix, '$');
    }

    public Name newName() {
        return Name.fromString("$" + (counter++) + "$");
    }
}

File: src/actors/scala/actors/remote/FreshNameCreator.scala
/*                     __                                               *\
**     ________ ___   / /  ___     Scala API                            **
**    / __/ __// _ | / /  / _ |    (c) 2005-2011, LAMP/EPFL             **
**  __\ \/ /__/ __ |/ /__/ __ |    http://scala-lang.org/               **
** /____/\___/_/ |_/____/_/ | |                                         **
**                          |/                                          **
\*                                                                      */


package scala.actors
package remote

object FreshNameCreator {

  protected var counter = 0
  protected val counters = new scala.collection.mutable.HashMap[String, Int]

  /**
   * Create a fresh name with the given prefix. It is guaranteed
   * that the returned name has never been returned by a previous
   * call to this function (provided the prefix does not end in a digit).
   */
  def newName(prefix: String): Symbol = {
    val count = counters.get(prefix) match {
      case Some(last) => last + 1
      case None => 0
    }
    counters.update(prefix, count)
    Symbol(prefix + count)
  }

  def newName(): Symbol = {
    counter += 1
    Symbol("$" + counter + "$")
  }
}
